syntax = "proto3";

package telemetry.v1;

option csharp_namespace = "Slipstream.Telemetry.V1";
option go_package       = "telemetry/v1;telemetryv1";
option java_multiple_files = true;
option java_package     = "ai.slipstream.telemetry.v1";
option java_outer_classname = "TelemetryProto";
option optimize_for     = SPEED;

message Vec3f   { float x = 1; float y = 2; float z = 3; }
message Wheel4f { float fl = 1; float fr = 2; float rl = 3; float rr = 4; }

message Physics {
  float speed_kmh           = 1;
  float gas                 = 2;
  float brake               = 3;
  float clutch              = 4;
  int32 gear                = 5;   // -1..7
  float steer               = 6;   // -1..1
  float engine_rpm          = 7;
  Wheel4f tyre_slip         = 8;
  Wheel4f tyre_load         = 9;
  Wheel4f tyre_temp_c       = 10;
  Wheel4f suspension_travel = 11;
  Vec3f acc_g               = 12;
  Vec3f velocity            = 13;  // m/s
  Vec3f angular_velocity    = 14;  // rad/s
}

message Graphics {
  int32 status               = 1;   // 0=AC_OFF, 1=AC_REPLAY, 2=AC_LIVE
  int32 completed_laps       = 2;
  int32 position             = 3;
  float i_current_time_s     = 4;
  float i_last_time_s        = 5;
  float i_best_time_s        = 6;
  float lap_time_s           = 7;
  float delta_lap_s          = 8;
  float normalized_car_pos   = 9;   // 0..1
}

message Static {
  string car_model           = 1;
  string track               = 2;
  string track_configuration = 3;
  int32  max_rpm             = 4;
}

message TelemetryFrame {
  uint64 monotonic_ns = 1;     // host steady_clock
  string game         = 2;     // "assetto_corsa"
  string session_id   = 3;     // uuid
  Physics physics     = 4;
  Graphics graphics   = 5;
  Static stat         = 6;
}

message StreamAck {
  enum AckStatus {
    ACK_STATUS_UNSPECIFIED = 0;
    ACK_OK = 1;
    ACK_INVALID = 2;
    ACK_BACKPRESSURE = 3;
    ACK_DROPPED = 4;
    ACK_ERROR = 5;
  }
  // latest timestamp acknowledged by server
  uint64 last_monotonic_ns = 1;
  AckStatus status = 2;
  string message = 3;
  uint32 buffered_frames = 4;
}

message SessionInfo {
  string session_id = 1;
  string driver_id = 2;
  string track_id = 3;
  string car_id = 4;
  uint64 started_at_ns = 5;
  uint64 closed_at_ns = 6;
  bool active = 7;
  map<string, string> tags = 8;
}

message CreateSessionRequest {
  string session_id = 1;
  string driver_id = 2;
  string track_id = 3;
  string car_id = 4;
  uint64 started_at_ns = 5;
  map<string, string> tags = 6;
}

message CreateSessionResponse {
  bool ok = 1;
  string message = 2;
  SessionInfo session = 3;
}

message CloseSessionRequest {
  string session_id = 1;
  uint64 closed_at_ns = 2;
}

message CloseSessionResponse {
  bool ok = 1;
  string message = 2;
  SessionInfo session = 3;
}

message QuerySessionTelemetryRequest {
  string session_id = 1;
  uint64 start_monotonic_ns = 2;
  uint64 end_monotonic_ns = 3;
  uint32 target_hz = 4;             // clamped to 10-20Hz
  bool normalize_distance_axis = 5; // 0..1 distance axis normalization
}

message QueryPoint {
  string session_id = 1;
  uint64 monotonic_ns = 2;
  uint64 timestamp_ns = 3;
  int32 lap = 4;
  float distance_norm = 5;
  float speed_kmh = 6;
  float throttle = 7;
  float brake = 8;
  float steer = 9;
  int32 gear = 10;
  float engine_rpm = 11;
  float wheel_speed_delta_kmh = 12;
  float lateral_g = 13;
}

message QuerySessionTelemetryResponse {
  repeated QueryPoint points = 1;
  uint32 source_count = 2;
  uint32 returned_count = 3;
  bool cached = 4;
}

message QueryLapSliceRequest {
  string session_id = 1;
  int32 lap = 2;
  float start_distance_norm = 3;
  bool has_start_distance_norm = 4;
  float end_distance_norm = 5;
  bool has_end_distance_norm = 6;
  uint32 target_hz = 7;
  bool normalize_distance_axis = 8;
}

message QueryLapSliceResponse {
  repeated QueryPoint points = 1;
  uint32 source_count = 2;
  uint32 returned_count = 3;
  bool cached = 4;
}

message OverlayPoint {
  float distance_norm = 1;
  float base_speed_kmh = 2;
  float compare_speed_kmh = 3;
  float speed_delta_kmh = 4;
  float base_throttle = 5;
  float compare_throttle = 6;
  float throttle_delta = 7;
  float base_brake = 8;
  float compare_brake = 9;
  float brake_delta = 10;
  float base_steer = 11;
  float compare_steer = 12;
  float steer_delta = 13;
}

message GetLapOverlayRequest {
  string base_session_id = 1;
  int32 base_lap = 2;
  string compare_session_id = 3;
  int32 compare_lap = 4;
  float start_distance_norm = 5;
  bool has_start_distance_norm = 6;
  float end_distance_norm = 7;
  bool has_end_distance_norm = 8;
  uint32 target_hz = 9;
}

message GetLapOverlayResponse {
  repeated OverlayPoint points = 1;
  uint32 base_source_count = 2;
  uint32 compare_source_count = 3;
  uint32 returned_count = 4;
  uint32 query_time_ms = 5;
  bool cached = 6;
}

message ListSessionsRequest {
  string driver_id = 1;
  string track_id = 2;
  string car_id = 3;
  bool active_only = 4;
  uint64 started_after_ns = 5;
  uint64 started_before_ns = 6;
  uint32 limit = 7;
}

message ListSessionsResponse {
  repeated SessionInfo sessions = 1;
  uint32 total_count = 2;
}

message LapSummary {
  int32 lap = 1;
  uint32 point_count = 2;
  float lap_time_s = 3;
  float avg_speed_kmh = 4;
  float peak_speed_kmh = 5;
}

message SessionSummary {
  SessionInfo session = 1;
  uint32 point_count = 2;
  uint32 lap_count = 3;
  LapSummary best_lap = 4;
  repeated LapSummary laps = 5;
}

message GetSessionSummaryRequest {
  string session_id = 1;
  bool include_laps = 2;
}

message GetSessionSummaryResponse {
  bool ok = 1;
  string message = 2;
  SessionSummary summary = 3;
}

service TelemetryIngest {
  rpc CreateSession(CreateSessionRequest) returns (CreateSessionResponse);
  rpc CloseSession(CloseSessionRequest) returns (CloseSessionResponse);
  rpc QuerySessionTelemetry(QuerySessionTelemetryRequest) returns (QuerySessionTelemetryResponse);
  rpc QueryLapSlice(QueryLapSliceRequest) returns (QueryLapSliceResponse);
  rpc GetLapOverlay(GetLapOverlayRequest) returns (GetLapOverlayResponse);
  rpc ListSessions(ListSessionsRequest) returns (ListSessionsResponse);
  rpc GetSessionSummary(GetSessionSummaryRequest) returns (GetSessionSummaryResponse);
  // Bidirectional stream; server may send acks or control hints
  rpc StreamFrames(stream TelemetryFrame) returns (stream StreamAck);
}
